apps/derive/elpi/derive.elpi

```diff
@@ -4,82 +4,118 @@
+validate-only _ _.
+
+pred known-option i:gref, i:list derive, i:prop.
+known-option T L (get-option X _) :-
+  if (std.mem! L (derive X _ _)) true
+     (coq.error "Derivation" X "unknown or not applicable to input" T).
+
+pred chain i:gref, i:list derive, o:list prop.
+chain _ [] [].
+chain T [derive Name _ _|FS] CL :- not(selected Name), !,
+  if-verbose (coq.say "Skipping derivation" Name "on" T "since the user did not select it"),
+  chain T FS CL.
+chain T [derive Name _ AlreadyDone|FS] CL :- (pi x\ stop x :- !, fail) => AlreadyDone, !,
+  if-verbose (coq.say "Skipping derivation" Name "on" T "since it has been already run"),
+  chain T FS CL.
+chain T [derive Name F _|FS] CL :- get-option "only" _, !, % request this one
+  if-verbose (coq.say "Derivation" Name "on" T),
+  @dropunivs! => std.time (F C) Time,
+  if-verbose (coq.say "Derivation" Name "on" T "took" Time),
+  C => chain T FS CS,
+  std.append C CS CL.
+chain T [derive Name F _|FS] CL :- % all are selected, we can fail
+  if-verbose (coq.say "Derivation" Name "on" T),
+  (pi x\ stop x :- !, fail) => @dropunivs! => std.time (F C) Time, !,
+  if-verbose (coq.say "Derivation" Name "on" T "took" Time),
+  C => chain T FS CS,
+  std.append C CS CL.
+chain T [derive F _ _|FS] CL :-
+  if-verbose (coq.say "Derivation" F "on" T "failed, continuing"),
+  chain T FS CL.
+
+pred toposort i:list derive, o:list derive.
+toposort L SL :-
+  std.findall (dep1 _ _) Deps,
+  topo L Deps SL.
+
+pred std.partition i:list A, i:(A -> prop), o:list A, o:list A.
+std.partition [] _ [] [].
+std.partition [X|XS] P [X|R] L :- P X, !, std.partition XS P R L.
+std.partition [X|XS] P R [X|L] :- std.partition XS P R L.
+
+pred not-a-src i:list prop, i:derive.
+not-a-src Deps (derive A _ _) :- not(std.mem! Deps (dep1 A _)).
+
+pred tgt-is-not-in i:list derive, i:prop.
+tgt-is-not-in [] _.
+tgt-is-not-in [derive Tgt _ _|_] (dep1 _ Tgt) :- !, fail.
+tgt-is-not-in [_|L] D :- tgt-is-not-in L D.
+
+pred topo i:list derive, i:list prop, o:list derive.
+topo [] _ [] :- !.
+topo L Deps SL :-
+  std.partition L (not-a-src Deps) LNoDeps Other,
+  if (LNoDeps = []) (coq.error "derive: no topological order:" L Deps) true,
+  std.filter Deps (tgt-is-not-in LNoDeps) NewDeps,
+  topo Other NewDeps SOther,
+  std.append LNoDeps SOther SL.
+
+pred export? i:prop, o:prop.
+export? (export M) (coq.env.export-module M).
+
+pred indt-or-const i:gref.
+indt-or-const (indt _).
+indt-or-const (const _).
+
+pred main i:gref, i:bool, o:list prop.
+main T P CL :- get-option "recursive" tt, !,
+  coq.env.dependencies T _ AllDeps,
+  coq.gref.set.elements AllDeps AllDepsL,
+  std.filter AllDepsL indt-or-const Deps,
+  main.aux Deps [] CL1,
+  CL1 => main1 T P CL2,
+  std.append CL1 CL2 CL.
+pred main.aux i:list gref, i:list prop, o:list prop.
+main.aux [] X X.
+main.aux [T|TS] Acc CL :-
+  /* (pi X\get-option "only" X :- !, fail) => */ Acc => main T tt CL1,
```

Comment:
Is it done on purpose, or is it some leftover?

Reply:
This was the thing I was explaining the other day during lunch, that recursively I was enabling all derivations.
It seems to work even if I don't do it, so I guess I was wrong and the commented code can go.

Reply:
Hum, if I try
```
Inductive t := | C : forall n, Nat.eqb n 3 = true -> t.

#[only(eqbOK),recursive] derive t.
```
I get
```
Derivation eqbOK unknown or not applicable to input const «Nat.eqb»
```
so probably it was useful to override the options passed to derive.

Reply:
fixed